File: Matchismo/ObjC/Matchismo/Matchismo/Model/PlayingCard.m
//
//  PlayingCard.m
//  Matchismo
//
//  Created by Bret Koppel on 8/21/14.
//  Copyright (c) 2014 Bret Koppel. All rights reserved.
//

#import "PlayingCard.h"

@implementation PlayingCard

// if you override BOTH getter and setter, must
// synthesize yourself.
@synthesize suit = _suit;

- (void)setSuit:(NSString *)suit{
    if ([[PlayingCard validSuits] containsObject:suit])
        _suit = suit;
}

- (NSString *)suit {
    return _suit ? _suit : @"?";
}

- (void)setRank:(NSUInteger)rank {
    if (rank <= [PlayingCard maxRank])
        _rank = rank;
}

- (NSString *)contents{
    NSArray *rankStrings = [PlayingCard rankStrings];
    return [rankStrings[self.rank] stringByAppendingString:self.suit];
}

+ (NSArray *)rankStrings {
    return @[@"?",@"A",@"2",@"3", @"4", @"5", @"6", @"7", @"8", @"9", @"10", @"J", @"Q", @"K"];
}

+ (NSArray *)validSuits {
    return @[@"♣︎",@"♦︎",@"♥︎",@"♠︎"];
}

+ (NSUInteger)maxRank {
    return [[self rankStrings] count] -1;
}

@end

File: Matchismo/Xamarin/Matchismo/Core/PlayingCard.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Core
{
	public class PlayingCard : Card
	{
		private string _suit = string.Empty;
		private int _rank;

		public override string Contents {
			get { return RankStrings [_rank] + _suit; }
			set { base.Contents = value; }
		}

		public string Suit
		{
			get{ return string.IsNullOrEmpty (_suit) ? "?" : _suit;}

			set
			{
				if (ValidSuits.Contains (value))
					_suit = value;
			}
		}

		public int Rank
		{
			get {return _rank;}
			set 
			{
				if (value <= MaxRank)
					_rank = value;
			}
		}

		public static IList<string> ValidSuits { get { return new[] {@"♣︎",@"♦︎",@"♥︎",@"♠︎"}; } }

		private static IList<string> RankStrings {
			get {
				return new[] {
					@"?",
					@"A",
					@"2",
					@"3",
					@"4",
					@"5",
					@"6",
					@"7",
					@"8",
					@"9",
					@"10",
					@"J",
					@"Q",
					@"K"
			};
			}
		}

		public static int MaxRank { get { return RankStrings.Count - 1; } }
	}
}

